File: scripts/algorithms/L/Largest Sum of Averages/Largest Sum of Averages.js
// Runtime: 124 ms (Top 33.33%) | Memory: 44.8 MB (Top 33.33%)
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var largestSumOfAverages = function(nums, k) {
    // set length
    const len = nums.length;
    // set sum by len fill
    const sum = new Array(len).fill(0);
    // set nums first to first of sum
    sum[0] = nums[0];

    // set every item of sum to the sum of the previous and the corresponding item of nums
    for (let i = 1; i < len; i++) {
        sum[i] = sum[i - 1] + nums[i];
    }

    // set dynamic programming
    const dp = new Array(k + 1).fill("").map(() => new Array(len).fill(0));

    // according to the meaning of the problem, set the value of dp
    for (let i = 0; i < len; i++) {
        dp[1][i] = sum[i] / (i + 1);
    }
    for (let i = 1; i <= k; i++) {
        dp[i][i - 1] = sum[i - 1];
    }
    for (let i = 2; i <= k; i++) {
        for (let j = i; j < len; j++) {
            for (let m = j - 1; m >= i - 2; m--) {
                dp[i][j] = Math.max(dp[i][j], dp[i - 1][m] + (sum[j] - sum[m]) / (j - m));
            }
        }
    }

    // result
    return dp[k][len - 1];
};

File: scripts/algorithms/L/Largest Sum of Averages/Largest Sum of Averages.cpp
// Runtime: 20 ms (Top 78.88%) | Memory: 8 MB (Top 44.62%)
class Solution {
public:
    double solve(vector<int>&nums, int index, int k, vector<vector<double>>&dp){
        if(index<0)
            return 0;
        if(k<=0)
            return -1e8;

        if(dp[index][k]!=-1)
            return dp[index][k];

        double s_sum = 0;
        double maxi = INT_MIN;
        int cnt = 1;
        for(int i=index;i>=0;i--){
            s_sum += nums[i];
            maxi = max(maxi, (s_sum/cnt) + solve(nums, i-1, k-1, dp));
            cnt++;
        }
        return dp[index][k] = maxi;
    }

    double largestSumOfAverages(vector<int>& nums, int k) {
        int n = nums.size();
        vector<vector<double>>dp(n, vector<double>(k+1, -1));
        return solve(nums, n-1, k, dp);
    }
};